### DISCUSSION TITLE
Skip/Optionally in ParserPrinter

### DISCUSSION THREAD

**Discussion by tikitu** [pointfreeco/swift-parsing#300] [Q&A] [answered]
I've got input that I don't want to round-trip completely: there are some optionals I'm happy to remove (e.g. an optional trailing comma in a list of items), and some variable input that's just plain irrelevant for my use case (think `PrefixUpTo("interesting stuff starts here")` and `Rest()` and similar). But I would still like to use ParserPrinters to get round-tripping for free for the parts of the input that I _do_ care about. What are my best options?

* `Skip` gives me `Generic struct 'ParsePrint' requires the types 'Substring' and '()' be equivalent`.
* `Optionally { "," }` still produces a `()?` because the printer wants to know whether the option was chosen or not.

I can get the commas with 

```swift
OneOf {
    ","
    Always(())
}
```
which seems awfully roundabout to me but does work. I don't know any way at all to drop the variable stuff, so for the moment I'm just carrying it around as unnecessary `Substring` arguments until I get to an `AnyConversion` where I can ignore it in `apply` and give `""` in `unapply`.

I'm sure there are better patterns than these, what am I missing?

**Comment by oskarek** [comment, marked as answer]
Sounds like you should be able to make use of the [`.printing(_:)` operator](https://github.com/pointfreeco/swift-parsing/blob/main/Sources/Parsing/ParserPrinters/Printing.swift) on `Parser`. It allows you to turn any `Parser` into a `ParserPrinter`, by explicitly telling it how to print.

Can be used like this:

```swift
let input = """
don't care don't care don't care
don't care really don't care
:123 don't care again
"""

struct MyParser: ParserPrinter {
  var body: some ParserPrinter<Substring, Int> {
    Skip { PrefixThrough(":") }.printing { _, _ in } // don't print anything
    Int.parser()
    Skip { Rest() }.printing { _, _ in } // don't print anything
  }
}

// parsing
let output = try MyParser().parse(input)
output // 123

// printing
let printedValue = try MyParser().print(100)
printedValue // "100"
```

**Reply by tikitu** [reply]
Beautiful, thanks!
